src/components/RangeSlider/components/DualThumb/DualThumb.tsx

```diff
@@ -0,0 +1,594 @@
+        steppedPercentageOfRange,
+        valueUpper,
+        min,
+        step,
+      );
+      this.setState(
+        {
+          valueLower: valueWithinBoundsLower,
+        },
+        () => {
+          this.handleChange();
+        },
+      );
+    }
+  }
+
+  @autobind
+  private handleMouseDownThumbUpper() {
+    if (this.thumbUpper.current) {
+      addEventListener(document, 'mousemove', this.handleMouseMoveThumbUpper);
+      addEventListener(
+        document,
+        'mouseup',
+        () => {
+          removeEventListener(
+            document,
+            'mousemove',
+            this.handleMouseMoveThumbUpper,
+          );
+        },
+        {once: true},
+      );
+    }
+  }
+
+  @autobind
+  private handleMouseMoveThumbUpper(event: MouseEvent) {
+    if (this.track.current) {
+      const {valueLower} = this.state;
+      const {min, max, step} = this.props;
+      const clientRect = this.track.current.getBoundingClientRect();
+
+      const relativeX = event.clientX - clientRect.left;
+
+      const percentageOfTrack = relativeX / (clientRect.width - THUMB_SIZE);
+      const percentageOfRange = percentageOfTrack * (max - min);
+
+      const steppedPercentageOfRange = roundToNearestStepValue(
+        percentageOfRange,
+        step,
+      );
+
+      const valueWithinBoundsUpper = keepValueWithinBoundsUpper(
+        steppedPercentageOfRange,
+        valueLower,
+        max,
+        step,
+      );
+      this.setState(
+        {
+          valueUpper: valueWithinBoundsUpper,
+        },
+        () => {
+          this.handleChange();
+        },
+      );
+    }
+  }
+
+  @autobind
+  private handleChange() {
+    const {onChange, id} = this.props;
+    const {valueLower, valueUpper} = this.state;
+
+    return onChange([valueLower, valueUpper], id);
+  }
+
+  @autobind
+  private handleTextFieldChangeLower(value: string) {
+    this.setState({valueLower: Number(value)});
+  }
+
+  @autobind
+  private handleTextFieldChangeUpper(value: string) {
+    this.setState({valueUpper: Number(value)});
+  }
+
+  @autobind
+  private handleTextFieldBlurLower() {
+    const {valueLower, valueUpper} = this.state;
+    const {step, min} = this.props;
+    const steppedValue = roundToNearestStepValue(valueLower, step);
+
+    const valueWithinBoundsLower = keepValueWithinBoundsLower(
+      steppedValue,
+      valueUpper,
+      min,
+      step,
+    );
+    this.setState(
+      {
+        valueLower: valueWithinBoundsLower,
+      },
+      () => {
+        this.handleChange();
+      },
+    );
+  }
+
+  @autobind
+  private handleTextFieldBlurUpper() {
+    const {step, max} = this.props;
+    const {valueUpper, valueLower} = this.state;
+    const steppedValue = roundToNearestStepValue(valueUpper, step);
+
+    const valueWithinBoundsUpper = keepValueWithinBoundsUpper(
+      steppedValue,
+      valueLower,
+      max,
+      step,
+    );
+    this.setState(
```

Comment:
We are only holding the track width and the value in state, yet are setting state in 10 different places. We should be able to pair that down to just a couple places where state is being set.